drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2724 @@
+
+        TD->HwTD.Padded[0] = 0;
+        TD->HwTD.Padded[1] = 0;
+
+        PrevTD->HwTD.gTD.NextTD = TD->PhysicalAddress;
+        PrevTD->NextHcdTD = (ULONG)TD;
+    }
+
+    if (TransferParameters->TransferFlags & USBD_SHORT_TRANSFER_OK)
+    {
+        PrevTD->HwTD.gTD.Control.BufferRounding = TRUE;
+        OhciTransfer->Flags |= OHCI_TRANSFER_FLAGS_SHORT_TRANSFER_OK;
+    }
+
+    TD->HwTD.gTD.CurrentBuffer = 0;
+    TD->HwTD.gTD.BufferEnd = 0;
+
+    TD->Flags |= OHCI_HCD_TD_FLAG_CONTROL_STATUS;
+    TD->TransferLen = 0;
+
+    TD->HwTD.gTD.Control.AsULONG = 0;
+
+    if ((TransferParameters->TransferFlags & USBD_TRANSFER_DIRECTION_IN) != 0)
+    {
+        TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_OUT;
+    }
+    else
+    {
+        TD->HwTD.gTD.Control.BufferRounding = TRUE;
+        TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_IN;
+    }
+
+    TD->HwTD.gTD.Control.DataToggle = OHCI_TD_DATA_TOGGLE_DATA1;
+    TD->HwTD.gTD.Control.ConditionCode = OHCI_TD_CONDITION_NOT_ACCESSED;
+
+    OhciTransfer->PendingTDs++;
+    OhciTransfer->ControlStatusTD = TD;
+
+    NextTD = OHCI_AllocateTD(OhciExtension, OhciEndpoint);
+
+    TD->HwTD.gTD.NextTD = NextTD->PhysicalAddress;
+    TD->NextHcdTD = (ULONG)NextTD;
+
+    NextTD->NextHcdTD = 0;
+    NextTD->HwTD.gTD.NextTD = 0;
+
+    OhciTransfer->NextTD = NextTD;
+    OhciEndpoint->HcdTailP = NextTD;
+
+    OhciEndpoint->HcdED->HwED.TailPointer = NextTD->PhysicalAddress;
+
+    OHCI_EnableList(OhciExtension, OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI 
+OHCI_BulkOrInterruptTransfer(IN POHCI_EXTENSION OhciExtension,
+                             IN POHCI_ENDPOINT OhciEndpoint,
+                             IN PUSBPORT_TRANSFER_PARAMETERS TransferParameters,
+                             IN POHCI_TRANSFER OhciTransfer,
+                             IN PUSBPORT_SCATTER_GATHER_LIST SGList)
+{
+    POHCI_HCD_TD TD;
+    POHCI_HCD_TD PrevTD;
+    ULONG TransferedLen;
+    ULONG MaxTDs;
+    ULONG MaxPacketSize;
+
+    DPRINT_OHCI("OHCI_BulkOrInterruptTransfer: ... \n");
+
+    MaxTDs = OHCI_RemainTDs(OhciExtension, OhciEndpoint);
+
+    if (SGList->SgElementCount > MaxTDs)
+    {
+        return MP_STATUS_FAILURE;
+    }
+
+    TD = OhciEndpoint->HcdTailP;
+
+    TransferedLen = 0;
+
+    do
+    {
+        TD->HwTD.gTD.Control.AsULONG = 0;
+        TD->HwTD.gTD.Control.DelayInterrupt = OHCI_TD_INTERRUPT_NONE;
+        TD->HwTD.gTD.Control.ConditionCode = OHCI_TD_CONDITION_NOT_ACCESSED;
+
+        if (TransferParameters->TransferFlags & USBD_TRANSFER_DIRECTION_IN)
+        {
+            TD->HwTD.gTD.Control.BufferRounding = 0;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_IN;
+        }
+        else
+        {
+            TD->HwTD.gTD.Control.BufferRounding = TRUE;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_OUT;
+        }
+
+        TD->HwTD.gTD.CurrentBuffer = 0;
```

Comment:
If these are pointers, i'd prefer `NULL`

Reply:
typedef struct _OHCI_TRANSFER_DESCRIPTOR { // must be aligned to a 16-byte boundary
...
  ULONG CurrentBuffer; // physical address of the next memory location
...
} OHCI_TRANSFER_DESCRIPTOR, *POHCI_TRANSFER_DESCRIPTOR;